### PR TITLE
refactor(core): validate element is known in JIT mode

### PR DESCRIPTION
**Pull request by cexbrayat**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the new behavior?

Renames, simplifies and documents the function in charge of validating if an element is known in JIT mode.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

This is a small change allowing to land #45479 with less noise

When this one lands, a similar PR will be opened to simplify `validateProperty`.

### CODE REVIEW

**Comment by cexbrayat** [comment]
Adding @AndrewKushnir as a reviewer as this is based on his suggestion in #45492

File: packages/core/src/render3/instructions/element.ts [hunk 1]
```diff
@@ -190,21 +191,31 @@ export function ɵɵelement(
...
   // If `schemas` is set to `null`, that's an indication that this Component was compiled in AOT
   // mode where this check happens at compile time. In JIT mode, `schemas` is always present and
   // defined as an array (as an empty array in case `schemas` field is not defined) and we should
   // execute the check below.
   if (schemas === null) return;
 
-  const tagName = tNode.value;
-
   // If the element matches any directive, it's considered as valid.
-  if (!hasDirectives && tagName !== null) {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/element.ts:206]: Just wondering why the `tagName` check is no longer needed (is this change intentional)?

File: packages/core/src/render3/instructions/element.ts [hunk 2]
```diff
@@ -190,21 +191,31 @@ export function ɵɵelement(
   return ɵɵelement;
 }
 
-function logUnknownElementError(
-    tView: TView, element: RElement, tNode: TNode, hasDirectives: boolean): void {
-  const schemas = tView.schemas;
-
+/**
+ * Validates that the element is known when running in JIT mode and logs an error if not.
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/element.ts:195]:
```suggestion
 * Validates that the element is known at runtime and produces
 * an error if it's not the case.
 * This check is relevant for JIT-compiled components (for AOT-compiled
 * ones this check happens at build time).
```

File: packages/core/src/render3/instructions/element.ts [hunk 3]
```diff
@@ -190,21 +191,31 @@ export function ɵɵelement(
...
+ *
+ * The element is considered known if either:
+ * - the component was compiled in AOT mode
+ * - the element matches any directive
+ * - the element is allowed by one of the schemas
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/element.ts:200]: I've dropped the AOT point (there is a comment on that above) and added a couple more items.
```suggestion
 * - it's a known HTML element
 * - it's a known custom element 
 * - the element matches any directive
 * - the element is allowed by one of the schemas
```

File: packages/core/src/render3/instructions/element.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/instructions/element.ts:206]: If we bring the check back, the `validateElementIsKnown` signature should also be updated to reflect that the `tagName` can be `null`.

File: packages/core/src/render3/instructions/shared.ts [hunk 4]
```diff
@@ -1108,9 +1109,12 @@ function validateProperty(
   return typeof Node === 'undefined' || Node === null || !(element instanceof Node);
 }
 
-export function matchingSchemas(tView: TView, tagName: string|null): boolean {
-  const schemas = tView.schemas;
-
+/**
+ * Returns true if one of the schemas allows the tag name provided.
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/shared.ts:1113]:
```suggestion
 * Returns true if the tag name is allowed by specified schemas.
```

**Review by AndrewKushnir — COMMENTED** [review]
@cexbrayat nice cleanup, thanks for the PR 👍 
I've left some minor comments, please let me know if you have any questions.

File: packages/core/src/render3/instructions/element.ts (hunk 1 above)
Comment (cexbrayat) [reply packages/core/src/render3/instructions/element.ts:206]: It looks like tagName can never be null: the function was previously called with a `TNode`, where the `tagName` is `string | null`, but was in fact only called with a `TElementNode`, where the tagName is `string`.

If you think it is safer to keep the check, I don't mind reverting the change.

Comment (AndrewKushnir) [reply packages/core/src/render3/instructions/element.ts:206]: Yeah, let's keep the check. The whole function is invoked in dev-mode only (tree-shaken away in prod), so that extra check would not affect the bundle.

**Review by AndrewKushnir — APPROVED** [review]
@cexbrayat thanks for the refactoring, the code looks great 👍 I've left one comment to bring back the `tagName` check, could you please take a look when you get a chance?

**Comment by cexbrayat** [comment]
@AndrewKushnir I re-added the null check on tagName: should be good to go for a last review.

**Review by AndrewKushnir — APPROVED** [review]
@cexbrayat thanks for the changes 👍  I'll kick off tests in Google's codebase shortly and will add the PR to the merge queue if there are no failing targets.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:439053450:BASE:439053449:1648927049871:13ae9680).

**Comment by AndrewKushnir** [comment]
@cexbrayat FYI the presubmit went well and I've added this PR to the merge queue. Thanks for the refactoring.
